include/pybind11/pybind11.h

```diff
@@ -2609,23 +2610,14 @@ class exception : public object {
 };
 
 PYBIND11_NAMESPACE_BEGIN(detail)
-// Returns a reference to a function-local static exception object used in the simple
-// register_exception approach below.  (It would be simpler to have the static local variable
-// directly in register_exception, but that makes clang <3.5 segfault - issue #1349).
-template <typename CppException>
-exception<CppException> &get_exception_object() {
-    static exception<CppException> ex;
```

Comment:
Does this default constructor call into CPython (and release/reacquire the GIL)?

The deadlock scenario can only arise if there's some non-trivial lock interaction in the initializer of the `static` variable.

I suppose this change would also deal with some end-of-program lifetime situations, though, e.g. if this object gets destroyed too early or too late.

Reply:
> Does this default constructor call into CPython (and release/reacquire the GIL)?

No, it just copies `nullptr` to `m_ptr` (https://github.com/pybind/pybind11/blob/3414c56b6c7c521d868c9a137ca2ace2e26b5b2e/include/pybind11/pytypes.h#L297).

> The deadlock scenario can only arise if there's some non-trivial lock interaction in the initializer of the `static` variable.

Yes, that was on my mind.

> I suppose this change would also deal with some end-of-program lifetime situations, though, e.g. if this object gets destroyed too early or too late.

That was the only reason I got into here.

I considered simply changing this to use `new` (with a boilerplate "intentional leak" comment), but then decided: the universally safe solution

1. is easy,
2. inexpensive,
3. expressive (call once is exactly what we want),
4. reads nicely,
5. future proof (in case the `exception<T>` implementation is changed),
6. and a great pattern for people to remember and follow,

let's use it!

Reply:
Yes, makes sense! Using the new facility just for the lifetime is certainly also a good move.